sunpy/data/manager/cache.py

```diff
@@ -0,0 +1,68 @@
+from sunpy.util.util import hash_file
+
+
+class Cache:
+    def __init__(self, downloader, storage):
+        self._downloader = downloader
+        self._storage = storage
+
+    def download(self, urls, redownload=False):
+        """
+        Downloads the files from the urls
+
+        Parameters
+        ----------
+        urls: List
+        list of urls
+        redownload: `bool`
+        Whether to skip cache and redownload
+        """
+        # TODO: Expiry time
+        # XXX: Expiry time cache level or download level?
+        if not redownload:
+            details = self._get_by_url(urls[0])
+            if details:
+                return details['file_path']
+
+        file_path, file_hash, url = self._download_and_hash(urls)
+
+        if not redownload:
+            self._storage.store({
+                'file_hash': file_hash,
+                'file_path': file_path,
+                'url': url,
+            })
+        return file_path
+
+    def get_by_hash(self, sha_hash):
+        """
+        Returns the details which is matched by hash if present in cache
+
+        Parameters
+        ----------
+        sha_hash: `str`
+        SHA-1 hash of the file
+        """
+        details = self._storage.find_by_key('file_hash', sha_hash)
+        return details
+
+    def _get_by_url(self, url):
```

Comment:
Is there a reason this is a private function?

Reply:
I didn't want to expose it. A cache (IMO) should handle this behind the scenes. But I ended up having to use it in data manager. So I guess it's up to discussion. 

Ping: @Cadair

Reply:
I mean there isn't an issue using the private function else where in SunPy. If you don't want a user to have to these functions normally, then it should be private.